private/nonlconAnalytic.m

```diff
@@ -75,32 +96,50 @@
 dc6_dq = reshape(dc6_dM, [1, 9]) * dM_dq;
 dc6_dx = [dc6_dq, 0];
 
-if useMaxNorm == false
-    c2 = (sum(g.^2,2)-gMax^2)'; %Nonlinear inequality constraint <= 0
 
-    dc2_dq = [bsxfun(@times, 2*firstDerivativeMatrix, g(:, 1))...
-              bsxfun(@times, 2*firstDerivativeMatrix, g(:, 2))...
-              bsxfun(@times, 2*firstDerivativeMatrix, g(:, 3))];
-    dc2_dx = [dc2_dq, zeros(N-1, 1)];
+% Constraint for ballistic motion encoding
+% For theory, please read/cite the following paper:
+% Szczepankiewicz F, Sjolund J, ... Westin C-F.
+% Motion-compensated gradient waveforms for tensor-valued
+% diffusion encoding by constrained numerical optimization
+% Magn Reson Med. 2020
+
+% Non-linear constraints for motion compensation
+if ~motionCompensation.linear && ~isempty(motionCompensation.order)
+    c7 = zeros(1, length(motionCompensation.order));
+    dc7_dx = zeros(length(motionCompensation.order), 3*N + 1);
+    
+    % dt is in ms; % Behaves better if calculation uses ms but requested tolerance
+    % has some strange units. Fixed this by rescaling the maxMagnitude by
+    % 1000^order.
```

Comment:
Does this problem persist even though the  'ScaleProblem' option in fmincon is enabled?

Reply:
I never had a chance to o systematic tests of this. 

I advise to keep my modification, and remove if necessary. I predict that it may become unstable otherwise.

Reply:
Ok let's leave it, after all this code has been validated.